### PR TITLE
AEA-1612 docs: add draft of 'acn-internals' page

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

docs: add draft of 'acn-internals' page

## Fixes

n/a

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes and CI passes too
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

n/a

### CODE REVIEW

File: docs/acn-internals.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+and the definition of the data structures involved is defined
+<a href="https://github.com/fetchai/agents-aea/blob/develop/libs/go/libp2p_node/acn/acn_message.proto" target="_blank">here</a>.
+
+## Messages
+
+### Agent Record
+
+An _Agent Record_ is a data structure containing information about an agent connected to the ACN.
+
+> TODO: list and explain fields types and purpose
```
Comment (DavidMinarsch) [inline docs/acn-internals.md:19]: This should be described as you point out. We also have this data structure currently present in two places which needs to be addressed.

File: docs/acn-internals.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+### ACN Message
+
+Entities in the ACN (i.e. either agents or peers) exchange _ACN messages_.
+An ACN message contains the following fields:
+
+- _version_: a string that stores the protocol version, following <a href="https://semver.org/" target="_blank">Semantic Versioning</a>.
```
Comment (DavidMinarsch) [inline docs/acn-internals.md:26]: See my branch https://github.com/fetchai/agents-aea/pull/2447, I am not sure there is any point in adding this. Either the protocol stays the same and is compatible, or we have a breaking change...

Comment (DavidMinarsch) [reply docs/acn-internals.md:26]: Also check out this branch for the full specification.

**Review by DavidMinarsch — COMMENTED** [review]
Good start

**Comment by DavidMinarsch** [comment]
Missing here are section from our internal Google Doc and the comments from Yuri on Slack.

File: docs/acn-internals.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+                note left of Agent: use DESERIALIZATION_ERROR
+            else error on decoding of Envelope payload
+                AgentApi->>Agent: Status(generic_error)
+                note left of Agent: use DESERIALIZATION_ERROR
+            else wrong payload
+                AgentApi->>Agent: Status(generic_error)
+                note left of Agent: use some custom error code
+            end
+        end
+        AgentApi ->> DHTPeer: RouteEnvelope
```
Comment (marcofavorito) [inline docs/acn-internals.md:167]: image

File: docs/acn-internals.md [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+## ACN with direct connection
+
+In the following sections, we describe the main three steps of the routing
+of an envelope through the ACN:
+
+- _ACN entrance_: when an envelope sent by an agent enters 
+  the peer-to-peer network via the peer the agent is connected to
+  i.e. agent-to-peer communication;
+- _ACN middle_: when an envelope gets routed through the peer-to-peer network,
```
Comment (marcofavorito) [inline docs/acn-internals.md:126]: better name than _ACN middle_?

File: docs/acn-internals.md [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+            else ack-timeout OR conn-error
+                note left of Agent: continue (Try to resend/reconnect)
+            else version not supported
+                DHTPeer->>Agent: Status(ERROR_UNSUPPORTED_VERSION)
+            else error on decoding of ACN message
+                DHTPeer->>Agent: Status(SERIALIZATION_ERROR)
+            else error on decoding of Envelope payload
+                DHTPeer->>Agent: Status(SERIALIZATION_ERROR)
+            else the payload cannot be handled
+                DHTPeer->>Agent: Status(SERIALIZATION_ERROR)
```
Comment (marcofavorito) [inline docs/acn-internals.md:162]: https://github.com/fetchai/agents-aea/pull/2447/files#r651544815

Comment (marcofavorito) [reply docs/acn-internals.md:162]: if the other comment will be addressed, this will change

File: docs/acn-internals.md [hunk 6]
```diff
@@ -0,0 +1,808 @@
...
+            note over Libp2p Node: announce agent<br/>address to<br/>other peers
+            Libp2p Node->>Libp2p Node: wait data from socket 
+            activate Libp2p Node
+            deactivate Libp2p Node
+        end
+</div>
+
+- The `send` method and the `receive` methods behave similarly to
+  the `send` and `receive` methods of the
+  <a href="#the-fetchaip2p_libp2p-connection" target="_blank">`p2p_libp2p connection</a>,
```
Comment (DavidMinarsch) [inline docs/acn-internals.md:752]:
```suggestion
  <a href="#the-fetchaip2p_libp2p-connection" target="_blank">`p2p_libp2p connection`</a>, 
```

**Review by DavidMinarsch — APPROVED** [review]
Great work @marcofavorito ! This will be very useful.

**Pull request merged**